fenix/app/src/main/java/org/mozilla/fenix/shopping/state/ReviewQualityCheckStore.kt

```diff
@@ -4,12 +4,70 @@
 
 package org.mozilla.fenix.shopping.state
 
+import kotlinx.coroutines.CoroutineScope
 import mozilla.components.lib.state.Store
 
 /**
  * Store for review quality check feature.
+ *
+ * @param reviewQualityCheckPreferences The [ReviewQualityCheckPreferences] instance to use.
+ * @param scope The [CoroutineScope] to use for launching coroutines.
  */
-class ReviewQualityCheckStore : Store<ReviewQualityCheckState, ReviewQualityCheckAction>(
+class ReviewQualityCheckStore(
+    reviewQualityCheckPreferences: ReviewQualityCheckPreferences,
+    scope: CoroutineScope,
+) : Store<ReviewQualityCheckState, ReviewQualityCheckAction>(
     initialState = ReviewQualityCheckState.Initial,
-    reducer = { _, _ -> ReviewQualityCheckState.Initial },
-)
+    middleware = listOf(
+        ReviewQualityCheckPreferencesMiddleware(reviewQualityCheckPreferences, scope),
+    ),
+    reducer = ::reducer,
+) {
+    init {
+        dispatch(ReviewQualityCheckAction.Init)
+    }
+}
+
+private fun reducer(
+    state: ReviewQualityCheckState,
+    action: ReviewQualityCheckAction,
+): ReviewQualityCheckState {
+    if (action is ReviewQualityCheckAction.UpdateAction) {
+        return mapStateForUpdateAction(state, action)
+    }
+
+    return state
+}
+
+private fun mapStateForUpdateAction(
+    state: ReviewQualityCheckState,
+    action: ReviewQualityCheckAction.UpdateAction,
+): ReviewQualityCheckState {
+    when (action) {
+        is ReviewQualityCheckAction.UpdateUserPreferences -> {
+            return if (action.hasUserOptedIn) {
+                if (state is ReviewQualityCheckState.OptedIn) {
+                    state.copy(productRecommendationsPreference = action.isProductRecommendationsEnabled)
+                } else {
+                    ReviewQualityCheckState.OptedIn(
+                        productRecommendationsPreference = action.isProductRecommendationsEnabled,
+                    )
+                }
+            } else {
+                ReviewQualityCheckState.NotOptedIn
+            }
+        }
+
+        ReviewQualityCheckAction.OptOut -> {
+            return ReviewQualityCheckState.NotOptedIn
+        }
+
+        ReviewQualityCheckAction.ToggleProductRecommendation -> {
+            return if (state is ReviewQualityCheckState.OptedIn) {
+                state.copy(productRecommendationsPreference = !state.productRecommendationsPreference)
+            } else {
+                state
+            }
```

Comment:
Will it be possible to toggle product recommendations while not opted in? If so, could we get a test describing that case? If not, could we get a comment describing that this isn't a reachable state?

Reply:
No, it won't be possible.